cogent/isa/Subtyping.thy

```diff
@@ -118,36 +118,22 @@ proof (induct t)
     by (fastforce simp add: list_all3_same list_all_iff intro!: type_lub_type_glb.intros)+
 next
   case (TRecord ts s)
-(*
-  assume ts_wellformed: "K \<turnstile> TRecord ts s wellformed"
-  {
-    fix n t t1 t2 b b1 b2
-    assume assms1:
-      "(n, t, b) \<in> set ts"
-      "(n, t1, b1) \<in> set ts"
-      "(n, t2, b2) \<in> set ts"
-    moreover have "t1 = t" "t2 = t" "b1 = b" "b2 = b"
-      using assms1 ts_wellformed
-        distinct_fst[where xs=ts and b="(p, q)" and b'="(p', q')" for p q p' q']
-      by fastforce+
-    ultimately have
-      "b1 = b" "b2 = b"
-      "K \<turnstile> t \<leftarrow> t1 \<squnion> t2 \<and> (b = inf b1 b2)"
-      "K \<turnstile> t \<leftarrow> t1 \<sqinter> t2 \<and> (b = sup b1 b2)"
-      using TRecord ts_wellformed
-      by auto
-  }
-  then show
-    "K \<turnstile> TRecord ts s \<leftarrow> TRecord ts s \<squnion> TRecord ts s"
-    "K \<turnstile> TRecord ts s \<leftarrow> TRecord ts s \<sqinter> TRecord ts s"
-    using ts_wellformed
-    by (auto intro!: type_lub_type_glb.intros)
-*)
-  then show
-    "K \<turnstile> TRecord ts s \<leftarrow> TRecord ts s \<squnion> TRecord ts s"
-    "K \<turnstile> TRecord ts s \<leftarrow> TRecord ts s \<sqinter> TRecord ts s"
-    sorry
-qed (force intro: type_lub_type_glb.intros)+
+  moreover assume ts_wellformed: "K \<turnstile> TRecord ts s wellformed"
+  ultimately show
+  "K \<turnstile> TRecord ts s \<leftarrow> TRecord ts s \<squnion> TRecord ts s"
+  "K \<turnstile> TRecord ts s \<leftarrow> TRecord ts s \<sqinter> TRecord ts s"
+     apply (rule_tac lub_trecord)
+           apply (metis (no_types, lifting) fsts.intros kinding_iff_wellformed(1) kinding_record_wellformed_elem kinding_simps(8) list_all3_same prod.collapse snds.intros)
```

Comment:
Here, the sledgehammer heuristics are using `kinding_iff_wellformed`, `kinding_simps(8)`, and `kinding_record_wellformed_elem` to prove that `?K ⊢ TRecord ?ts ?s wellformed ⟹ (?name, ?t, ?taken) ∈ set ?ts ⟹ ?K ⊢ ?t wellformed`.

The simplifier should be able to work this out, but it's struggling to do so in the context of the bigger lemma. At this point, given sledgehammer has done this a few times, it's probably a good idea to break it out into its own lemma in `Cogent.thy` (also include one for TSum).

Reply:
How did you that sledgehammer heuristics was proving `?K ⊢ TRecord ?ts ?s wellformed ⟹ (?name, ?t, ?taken) ∈ set ?ts ⟹ ?K ⊢ ?t wellformed`? Did you work it out yourself?

Reply:
I worked it out from what it needed to prove; I'd seen it do this earlier, so it was easy to guess. You can tell from the fact the lemma doesn't mention kinding, only wellformed. The kinding judgement is built on top of wellformedness, so it's usually odd if you need to produce a wellformedness proof from another wellformedness proof, and go through a kinding judgement.